### PR TITLE
Feature/ Adds xml and csv parsing

### PR DESCRIPTION
**Pull request by alexbourret**
Origin card: [SC-104769](https://app.shortcut.com/dataiku/story/104769/handle-apis-returning-csv-or-xml)

### CODE REVIEW

File: python-connectors/api-connect_dataset/connector.py [hunk 1]
```diff
@@ -49,9 +52,14 @@ def generate_rows(self, dataset_schema=None, dataset_partitioning=None,
                 record_count += len(data)
                 for row in data:
                     yield self.format_output(row, metadata)
-            else:
+            elif isinstance(data, dict):
                 record_count += 1
                 yield self.format_output(data, metadata)
+            else:
+                data = decode_csv_data(data)
```
Comment (vleocmach) [inline python-connectors/api-connect_dataset/connector.py:59]: looks like csv format decoding is applied as a fallback when nothing else could be applied. Can't we handle this the same way as xml format, checking response header 'text/csv' ?

File: python-lib/rest_api_recipe_session.py [hunk 2]
```diff
@@ -117,7 +118,13 @@ def retrieve_next_page(self, is_raw_output):
                         base_row.update(parse_keys_for_json(row))
                         base_row.update(self.initial_parameter_columns)
                         page_rows.append(base_row)
-
+                else:
+                    json_response = decode_csv_data(json_response)
```
Comment (vleocmach) [inline python-lib/rest_api_recipe_session.py:122]: same remark as for dataset#generate_rows logic; csv decoding looks like a fallback when nothing works. I'm wondering if we can handle this format prior to that, testing response header content-type as for xml.

**Review by vleocmach — COMMENTED** [review]
Tested with csv and xml apis for both plugin Datasets and Recipes; works well :)  !
I have some comments on the code to try to make it a bit clearer though.
But those may be linked with my newbie plugin understanding

File: python-lib/dku_utils.py [hunk 3]
```diff
@@ -96,3 +100,37 @@ def extract_key_using_json_path(json_dictionary, json_path):
         return res
     else:
         return None
+
+
+def is_reponse_xml(response):
+    content_types = response.headers.get("Content-Type", "").split(";")
+    for content_type in content_types:
+        if content_type in ["text/xml", "application/soap+xml", "text/plain", "application/xml"]:
```
[the remark this answers is not in the document]
Comment (alexbourret) [reply python-lib/dku_utils.py]: Removed in [838e8ff](https://github.com/dataiku/dss-plugin-api-connect/pull/33/commits/838e8ff4be0a0e6fbca6e93943312142611775f5)

File: python-lib/rest_api_client.py [hunk 4]
```diff
@@ -236,3 +227,32 @@ def enforce_throttling(self):
 
     def get_metadata(self):
         return self.metadata
+
+    def get_json_from_response(self, response, can_raise_exeption=True):
+        json_response = None
+        if is_reponse_xml(response):
+            logger.info("XML reply detected, converting to JSON")
+            json_response = xml_to_json(response.content)
+        else:
```
[the remark this answers is not in the document]
Comment (alexbourret) [reply python-lib/rest_api_client.py:236]: The approach here is to try to decode and raise only if nothing works. The plugin is meant to support as many format as possible, even from poorly coded APIs.

File: python-lib/dku_utils.py [hunk 5]
```diff
@@ -96,3 +100,37 @@ def extract_key_using_json_path(json_dictionary, json_path):
...
+
+
+def decode_csv_data(data):
+    import csv
+    import io
+    json_data = None
+    if isinstance(data, bytes):
+        data = data.decode("utf-8")
+    try:
+        reader = csv.DictReader(io.StringIO(data))
```
[the remark this answers is not in the document]
Comment (alexbourret) [reply python-lib/dku_utils.py:131]: The python kernel is killed at the end of the job.

File: python-connectors/api-connect_dataset/connector.py (hunk 1 above)
Comment (alexbourret) [reply python-connectors/api-connect_dataset/connector.py:59]: This is indeed a last chance fallback, for cases where the path to data array points to a string. It is there to handle csv encapsulated in xml (or json for that matter). Therefore the header can be anything.

File: python-lib/rest_api_recipe_session.py (hunk 2 above)
Comment (alexbourret) [reply python-lib/rest_api_recipe_session.py:122]: The recipe's behavior is modeled on the connector's.

File: python-connectors/api-connect_dataset/connector.py (hunk 1 above)
Comment (vleocmach) [reply python-connectors/api-connect_dataset/connector.py:59]: oh ok I didn't have this in mind, thanks :)

**Review by vleocmach — APPROVED** [review]
Thanks for the details, LGTM :)

**Pull request merged**